# Проект VK

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px

In [2]:
data = pd.read_excel('/Users/alexeystarodubtsev/Downloads/Analytic_test.xlsx', usecols = 'B:Y', skiprows=1)

In [3]:
data

HID  active_Red  has_Green_id  \
0         00003c7fb5e57ce673f1801f3679f5c25         0.0           0.0   
1         00006b602533083e3bebfcbff08613931         1.0           0.0   
2          0000736c3f76339196cb1cc759b8363b         1.0           0.0   
3          00009bdcc5307b38efde8690b8bd658e         0.0           0.0   
4         0000bc3186b7b31d9b35d3fbbb250b9f8         1.0           0.0   
...                                     ...         ...           ...   
99995  32555ec33d762588db3253b063525f03b58f         0.0           0.0   
99996  3255600b525b2598b37775fb3b25b3397bbb         1.0           0.0   
99997     325563dbbb35c393be38c5c79373ed180         1.0           0.0   
99998   3255650825570bcc93bc9b076b0bfbb25e6         0.0           0.0   
99999     325566ec18c0db59083f5f6bb1b5f93d5         0.0           0.0   

       has_Black_id  has_Blue_id Blue_activity_count  Blue_gmv_total  \
0               0.0          1.0                 2.0             0.0   
1               0.0          0.0                 0.0             0.0   
2               0.0          0.0                 0.0             0.0   
3               0.0          0.0                 0.0             0.0   
4               0.0          0.0                 0.0             0.0   
...             ...          ...                 ...             ...   
99995           0.0          0.0                 0.0             0.0   
99996           0.0          0.0                 0.0             0.0   
99997           0.0          0.0                 0.0             0.0   
99998           0.0          0.0                 0.0             0.0   
99999           0.0          0.0                 0.0             0.0   

       Green_gvm_total Green_gvm_order_count  Black_gvm_total  ...  \
0                  0.0                   0.0              0.0  ...   
1                  0.0                   0.0              0.0  ...   
2                  0.0                   0.0              0.0  ...   
3                  0.0                   0.0              0.0  ...   
4                  0.0                   0.0              0.0  ...   
...                ...                   ...              ...  ...   
99995              0.0                   0.0              0.0  ...   
99996              0.0                   0.0              0.0  ...   
99997              0.0                   0.0              0.0  ...   
99998              0.0                   0.0              0.0  ...   
99999              0.0                   0.0              0.0  ...   

      Blue_android  Red_apple  Green_apple  Black_apple  Blue_apple  is_msk  \
0              0.0        0.0          0.0          0.0         0.0     0.0   
1              0.0        0.0          0.0          0.0         0.0     0.0   
2              0.0        1.0          0.0          1.0         0.0     0.0   
3              0.0        0.0          0.0          0.0         0.0     0.0   
4              0.0        1.0          0.0          0.0         0.0     0.0   
...            ...        ...          ...          ...         ...     ...   
99995          0.0        0.0          0.0          0.0         0.0     1.0   
99996          0.0        0.0          0.0          0.0         0.0     0.0   
99997          0.0        0.0          0.0          0.0         0.0     0.0   
99998          0.0        0.0          0.0          0.0         0.0     0.0   
99999          0.0        0.0          0.0          0.0         0.0     0.0   

       Red_web  Green_web  Black_web  Blue_web  
0          0.0        0.0        0.0       0.0  
1          1.0        0.0        0.0       0.0  
2          0.0        0.0        0.0       0.0  
3          1.0        0.0        0.0       0.0  
4          0.0        0.0        0.0       0.0  
...        ...        ...        ...       ...  
99995      1.0        0.0        0.0       0.0  
99996      1.0        0.0        0.0       0.0  
99997      0.0        0.0        0.0       0.0  
99998      1.0    

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 24 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   HID                    100000 non-null  object 
 1   active_Red             100000 non-null  float64
 2   has_Green_id           100000 non-null  float64
 3   has_Black_id           100000 non-null  float64
 4   has_Blue_id            100000 non-null  float64
 5   Blue_activity_count    100000 non-null  object 
 6   Blue_gmv_total         99999 non-null   float64
 7   Green_gvm_total        100000 non-null  float64
 8   Green_gvm_order_count  100000 non-null  object 
 9   Black_gvm_total        100000 non-null  float64
 10  Black_gvm_order_count  100000 non-null  object 
 11  Red_android            100000 non-null  float64
 12  Green_android          100000 non-null  float64
 13  Black_android          100000 non-null  float64
 14  Blue_android           100000 non-nul

In [5]:
print(f'Количество уникальных HID:', len(data['HID'].unique()))

Количество уникальных HID: 100000


## Предобработка данных

In [6]:
data = data.drop_duplicates().reset_index(drop=True)
data = data.dropna().reset_index(drop=True)

data['Blue_activity_count'] = pd.to_numeric(data['Blue_activity_count'], errors='coerce')
data['Green_gvm_order_count'] = pd.to_numeric(data['Green_gvm_order_count'], errors='coerce')
data['Black_gvm_order_count'] = pd.to_numeric(data['Black_gvm_order_count'], errors='coerce')

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   HID                    99999 non-null  object 
 1   active_Red             99999 non-null  float64
 2   has_Green_id           99999 non-null  float64
 3   has_Black_id           99999 non-null  float64
 4   has_Blue_id            99999 non-null  float64
 5   Blue_activity_count    99992 non-null  float64
 6   Blue_gmv_total         99999 non-null  float64
 7   Green_gvm_total        99999 non-null  float64
 8   Green_gvm_order_count  99996 non-null  float64
 9   Black_gvm_total        99999 non-null  float64
 10  Black_gvm_order_count  99996 non-null  float64
 11  Red_android            99999 non-null  float64
 12  Green_android          99999 non-null  float64
 13  Black_android          99999 non-null  float64
 14  Blue_android           99999 non-null  float64
 15  Re

## Количество зарегистрированных пользователей по компаниям

In [16]:
company_cnt_users = data[['active_Red', 'has_Green_id', 'has_Black_id', 'has_Blue_id']].sum().sort_values(ascending = False)
print("\nКоличество зарегистрированных пользователей по компаниям:")
print(company_cnt_users)


Количество зарегистрированных пользователей по компаниям:
active_Red      55874.0
has_Blue_id     13798.0
has_Black_id     1084.0
has_Green_id      889.0
dtype: float64


In [17]:
company_cnt_users.rename(index={'active_Red': 'Red', 'has_Green_id': 'Green', 'has_Black_id': 'Black', 'has_Blue_id': 'Blue'}, inplace=True)

fig_company_cnt_users = px.bar(company_cnt_users, x=company_cnt_users.index, y=company_cnt_users.values, 
                                 title='Количество пользователей по компаниям')

fig_company_cnt_users.update_layout(xaxis_title='Комапния',
                   yaxis_title='Количество пользователей')

fig_company_cnt_users.show()

<div style="border:solid black 2px; padding: 20px">

Наибольшее число пользователей у компании **Red** (55874 зарегистрированных пользоватлей). На втором месте идёт компания **Blue** (13801 зарегистрированных пользоватлей). Замыкают список: **Black** (1084 зарегистрированных пользоватлей) и **Green** (890 зарегистрированных пользоватлей).

## Общая сумма заказов по компаниям

In [18]:
company_gmv_total = data[['Blue_gmv_total', 'Green_gvm_total', 'Black_gvm_total']].sum().sort_values(ascending = False)
print("\nОбщая сумма заказов по компаниям:")
print(company_gmv_total)


Общая сумма заказов по компаниям:
Black_gvm_total    9487308.0
Green_gvm_total    5992600.0
Blue_gmv_total      246114.0
dtype: float64


In [19]:
company_gmv_total.rename(index={'Green_gvm_total': 'Green', 'Black_gvm_total': 'Black', 'Blue_gmv_total': 'Blue'}, inplace=True)

fig_company_gmv_total = px.bar(company_gmv_total, x=company_gmv_total.index, y=company_gmv_total.values, 
                                 title='Общая сумма заказов по компаниям')

fig_company_gmv_total.update_layout(xaxis_title='Комапния',
                   yaxis_title='Общая сумма заказов')

fig_company_gmv_total.show()

<div style="border:solid black 2px; padding: 20px">

Наибольшая общая сумма заказов у компании **Black** (9487308). На втором месте идёт компания **Green** (5992600). Замыкает список: **Blue** (246114).

## Количество заказов по компаниям

In [20]:
company_order_count = data[['Green_gvm_order_count', 'Black_gvm_order_count']].sum().sort_values(ascending = False)
print("\nКоличество заказов по компаниям:")
print(company_order_count)


Количество заказов по компаниям:
Green_gvm_order_count    7698.0
Black_gvm_order_count    4385.0
dtype: float64


In [21]:
company_order_count.rename(index={'Green_gvm_order_count': 'Green', 'Black_gvm_order_count': 'Black'}, inplace=True)

fig_company_order_count = px.bar(company_order_count, x=company_order_count.index, y=company_order_count.values, 
                                 title='Общее количество заказов по компаниям')

fig_company_order_count.update_layout(xaxis_title='Комапния',
                   yaxis_title='Количество заказов')

fig_company_order_count.show()

<div style="border:solid black 2px; padding: 20px">
    
Исходя из предоставленных данных (отсутствие данных по Red и Blue), наибольшее количество успешных заказов у компании **Green** (7698). У **Black** значительно меньше успешных заказов (4385).

In [14]:
company_cnt_users

Red      55874.0
Green      889.0
Black     1084.0
Blue     13798.0
dtype: float64

## Активность пользователей компании Blue

In [22]:
# Считаем среднее число активных действий на общее число зарегестрированных пользователей компании Blue
blue_avg_activities =  data['Blue_activity_count'].sum() / data['has_Blue_id'].sum()

print(f'Среднее число активных действий на количество зарегестрированных пользователей компании Blue:', 
      round(blue_avg_activities, 2))

Среднее число активных действий на количество зарегестрированных пользователей компании Blue: 1.52


In [23]:
# Считаем количество пользователей компании Blue, которые совершили хотя бы 1 активное действие
blue_active_users_count = len(data[data['Blue_activity_count'] > 0]['HID'].unique())

# Считаем долю пользователей компании Blue, которые совершили хотя бы 1 активное действие
blue_active_users_share = blue_active_users_count / data['has_Blue_id'].sum()

# Выводим результат
print("Процент пользователей компании Blue, которые совершили хотя бы 1 активное действие: {:.2%}".format(blue_active_users_share))

Процент пользователей компании Blue, которые совершили хотя бы 1 активное действие: 28.12%


## Установленные приложения по платформам

### Среди всех компаний и устройств

In [24]:
platforms = data[['Red_android', 'Green_android', 'Black_android', 'Blue_android', 'Red_apple', 'Green_apple', 'Black_apple', 'Blue_apple', 'Red_web', 'Green_web', 'Black_web', 'Blue_web']].sum()
print("\nРаспределение установленных приложений по платформам:")
platforms = platforms.sort_values(ascending = False)
platforms


Распределение установленных приложений по платформам:


Red_web          67108.0
Red_android      32133.0
Red_apple        10874.0
Blue_web         10640.0
Blue_android      8047.0
Blue_apple        4646.0
Black_apple       1782.0
Black_web         1461.0
Black_android     1077.0
Green_apple        616.0
Green_android      546.0
Green_web          198.0
dtype: float64

In [25]:
fig_platforms = px.bar(platforms, x=platforms.index, y=platforms.values, 
                                 title='Общее число установленных приложений по компаниям и платформам')

fig_platforms.update_layout(xaxis_title='Компания и устройство',
                   yaxis_title='Количество заказов')

fig_platforms.show()

### По компаниям

#### Red

In [26]:
platforms_red = platforms[['Red_web', 'Red_android', 'Red_apple']]
print("\nЧисло установленных приложений пользователями Red:")
print(platforms_red.sort_values(ascending = False))


Число установленных приложений пользователями Red:
Red_web        67108.0
Red_android    32133.0
Red_apple      10874.0
dtype: float64


In [27]:
fig_platforms_red = px.bar(platforms_red, x=platforms_red.index, y=platforms_red.values, 
                                 title='Общее число установленных приложений по платформам пользователями Red')

fig_platforms_red.update_layout(xaxis_title='Устройство',
                   yaxis_title='Количество установленных приложений')

fig_platforms_red.show()

<div style="border:solid black 2px; padding: 20px">
    
Наибольшее число пользователей Red установили приложения на **персональные компютеры** (67108 установок). Далее по установкам идут приложения на **Android** (32133 установок) и **Apple** (10874 установок).

#### Blue

In [28]:
platforms_blue = platforms[['Blue_web', 'Blue_android', 'Blue_apple']]
print("\nЧисло установленных приложений пользователями Blue:")
print(platforms_blue.sort_values(ascending = False))


Число установленных приложений пользователями Blue:
Blue_web        10640.0
Blue_android     8047.0
Blue_apple       4646.0
dtype: float64


In [29]:
fig_platforms_blue = px.bar(platforms_blue, x=platforms_blue.index, y=platforms_blue.values, 
                                 title='Общее число установленных приложений по платформам пользователями Blue')

fig_platforms_blue.update_layout(xaxis_title='Устройство',
                   yaxis_title='Количество установленных приложений')

fig_platforms_blue.show()

<div style="border:solid black 2px; padding: 20px">
    
Наибольшее число пользователей Blue установили приложения на **персональные компютеры** (10640 установок). Далее по установкам идут приложения на **Android** (8047 установок) и **Apple** (4646 установок).

#### Black

In [30]:
platforms_black = platforms[['Black_web', 'Black_android', 'Black_apple']]
print("\nЧисло установленных приложений пользователями Black:")
platforms_black = platforms_black.sort_values(ascending = False)
platforms_black


Число установленных приложений пользователями Black:


Black_apple      1782.0
Black_web        1461.0
Black_android    1077.0
dtype: float64

In [31]:
fig_platforms_black = px.bar(platforms_black, x=platforms_black.index, y=platforms_black.values, 
                                 title='Общее число установленных приложений по платформам пользователями Black')

fig_platforms_black.update_layout(xaxis_title='Устройство',
                   yaxis_title='Количество установленных приложений')

fig_platforms_black.show()

<div style="border:solid black 2px; padding: 20px">
    
Наибольшее число пользователей Black установили приложения на **Apple** (1782 установок). Далее по установкам идут приложения на **персональные компьютеры** (1461 установок) и **Android** (1077 установок).

#### Green

In [32]:
platforms_green = platforms[['Green_web', 'Green_android', 'Green_apple']]
print("\nЧисло установленных приложений пользователями Green:")
platforms_green = platforms_green.sort_values(ascending = False)
platforms_green


Число установленных приложений пользователями Green:


Green_apple      616.0
Green_android    546.0
Green_web        198.0
dtype: float64

In [33]:
fig_platforms_green = px.bar(platforms_green, x=platforms_green.index, y=platforms_green.values, 
                                 title='Общее число установленных приложений по платформам пользователями Green')

fig_platforms_green.update_layout(xaxis_title='Устройство',
                   yaxis_title='Количество установленных приложений')

fig_platforms_green.show()

<div style="border:solid black 2px; padding: 20px">
    
Наибольшее число пользователей Green установили приложения на **Apple** (616 установок). Далее по установкам идут приложения на **персональные компьютеры** (546 установок) и **Android** (198 установок).

## Количество пользователей из Москвы

### Red

In [34]:
red_msk_users = data[(data['active_Red'] == 1) & (data['is_msk'] == 1)]
print(f'Абсолютное количество пользователей из Москвы', len(red_msk_users))
print("Процент пользователей компании Red из Москвы: {:.2%}".format(len(red_msk_users)/data['active_Red'].sum()))

Абсолютное количество пользователей из Москвы 6672
Процент пользователей компании Red из Москвы: 11.94%


### Blue

In [35]:
blue_msk_users = data[(data['has_Blue_id'] == 1) & (data['is_msk'] == 1)]
print(f'Абсолютное количество пользователей из Москвы', len(blue_msk_users))
print("Процент пользователей компании Blue из Москвы: {:.2%}".format(len(blue_msk_users)/data['has_Blue_id'].sum()))

Абсолютное количество пользователей из Москвы 2033
Процент пользователей компании Blue из Москвы: 14.73%


### Black

In [36]:
black_msk_users = data[(data['has_Black_id'] == 1) & (data['is_msk'] == 1)]
print(f'Абсолютное количество пользователей из Москвы', len(black_msk_users))
print("Процент пользователей компании Black из Москвы: {:.2%}".format(len(black_msk_users)/data['has_Black_id'].sum()))

Абсолютное количество пользователей из Москвы 388
Процент пользователей компании Black из Москвы: 35.79%


### Green

In [37]:
green_msk_users = data[(data['has_Green_id'] == 1) & (data['is_msk'] == 1)]
print(f'Абсолютное количество пользователей из Москвы', len(green_msk_users))
print("Процент пользователей компании Green из Москвы: {:.2%}".format(len(green_msk_users)/data['has_Green_id'].sum()))

Абсолютное количество пользователей из Москвы 674
Процент пользователей компании Green из Москвы: 75.82%


### Общий вывод 

<div style="border:solid black 2px; padding: 20px">

При изучении предоставленных данных о компаниях были выделены ****следующие особенности****:

**Red**

- 55874 зарегистрированных пользоватлей, 
- Абсолютное количество пользователей из Москвы 6672
- Процент пользователей компании Red из Москвы: 11.94%
- Наибольшее число пользователей Red установили приложения на **персональные компютеры** (67108 установок). Далее по установкам идут приложения на **Android** (32133 установок) и **Apple** (10874 установок).

Компания предоставляет достаточно популярные среди пользователей услуги. 11.94% пользователей из Москвы говорит, что компания работает не только в столице. Виртуальный сервис компании является универсальным и мультиплатформенным.

**Вlue**

- 13798 зарегистрированных пользоватлей,
- Абсолютное количество пользователей из Москвы 2033
- Процент пользователей компании Вlue из Москвы: 14.73%
- Общая сумма заказов у компании: 246114
- Среднее число активных действий на количество зарегестрированных пользователей компании Blue: 1.52
- Процент пользователей компании Blue, которые совершили хотя бы 1 активное действие: 28.12%
- Наибольшее число пользователей Blue установили приложения на **персональные компютеры** (10640 установок). Далее по установкам идут приложения на **Android** (8047 установок) и **Apple** (4646 установок).

Компания является второй по количеству зарегистрированных пользователей. По сумме заказов имеет наименьшее значение из всех (246114). При этом можно заметить низкую активность пользователей. Только 28.12% совершили хотя бы 1 активное действие. 14.73% пользователей из Москвы говорит, что компания работает не только в столице. Виртуальный сервис компании является универсальным и мультиплатформенным.

**Black**

- 1084 зарегистрированных пользоватлей,
- Абсолютное количество пользователей из Москвы 2033
- Процент пользователей компании Black из Москвы: 14.73%
- Общая сумма заказов: 9487308 
- Количество заказов: 4385
- Абсолютное количество пользователей из Москвы 388
- Процент пользователей компании Black из Москвы: 35.79%
- Наибольшее число пользователей Black установили приложения на **Apple** (1782 установок). Далее по установкам идут приложения на **персональные компьютеры** (1461 установок) и **Android** (1077 установок).

У компании Black значительно меньше пользователей, чем у двух предыдущих. При этом компания имеет наибольшую сумму заказов при относительно небольшом количестве заказов. Компания предоставляет дорогие услуги для своей небольшой аудитории, которая пользуется премиальными устройствами (Apple) и живет в городах (35.79% пользователей из Москвы)

**Green**

- 889 зарегистрированных пользоватлей, 
- Абсолютное количество пользователей из Москвы 2033
- Процент пользователей компании Green из Москвы: 14.73%
- Общая сумма заказов: 5992600
- Количество заказов: 7698
- Абсолютное количество пользователей из Москвы 674
- Процент пользователей компании Green из Москвы: 75.82%
- Наибольшее число пользователей Green установили приложения на **Apple** (616 установок). Далее по установкам идут приложения на **персональные компьютеры** (546 установок) и **Android** (198 установок).

Компания замыкает список по количеству зарегистрированных пользователей. Она имеет набольшее количество заказов, но у нее значительно меньше общая сумма заказов, в отличие от Black. Большинство пользователей пользуется 
премиальными устройствами (Apple) и живет в Москве (75.82% пользователей из Москвы).